src/adaptation/precond.jl

```diff
@@ -2,389 +2,227 @@
+    is_update::Bool=true
+) = nothing
 
-# https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
-function get_var(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    n, M = T(wv.n), wv.M
-    @assert n >= 2 "Cannot get covariance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M .+ T(1e-3) * (5 / (n + 5))
-end
+## Diagonal mass matrix adaptor
 
-## Covariance estimator
+abstract type VarEstimator{T} <: MassMatrixAdaptor end
 
-abstract type CovEstimator{T} end
+getM⁻¹(ve::VarEstimator) = ve.var
 
-# NOTE: This naive covariance estimator is used only in testing.
-mutable struct NaiveCov{T<:AbstractVector{<:AbstractVector{<:AbstractFloat}}} <: CovEstimator{T}
+Base.string(ve::VarEstimator) = string(getM⁻¹(ve))
+
+function adapt!(
+    ve::VarEstimator,
+    θ::AbstractVecOrMat{<:AbstractFloat},
+    α::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+)
+    resize!(ve, θ)
+    push!(ve, θ)
+    if ve.n >= ve.n_min && is_update
+        ve.var .= getest(ve)
+    end
+end
+
+# NOTE: this naive variance estimator is used only in testing
+mutable struct NaiveVar{E<:AbstractVecOrMat{<:AbstractFloat},T<:AbstractVector{E}} <: VarEstimator{T}
     n :: Int
     S :: T
 end
 
-NaiveCov(::Type{T}, sz::Union{Tuple{Int}}) where {T<:AbstractFloat} = NaiveCov(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int,Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Matrix{T}}())
 
-function add_sample!(nc::NaiveCov, s::AbstractVector)
-    nc.n += 1
-    push!(nc.S, s)
-end
+NaiveVar(sz::Tuple{Vararg{Int}}; kwargs...) = NaiveVar(Float64, sz; kwargs...)
 
-function get_cov(nc::NaiveCov)
-    @assert nc.n >= 2 "Cannot get covariance with only one sample"
-    return Statistics.cov(nc.S)
+function Base.push!(nv::NaiveVar, s::AbstractVecOrMat)
+    nv.n += 1
+    push!(nv.S, s)
 end
 
-function reset!(nc::NaiveCov{T}) where {T}
-    nc.n = 0
-    nc.S = T()
+function reset!(nv::NaiveVar{T}) where {T}
+    nv.n = 0
+    nv.S = T()
 end
 
-# Ref: https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_covar_estimator.hpp
-mutable struct WelfordCov{T<:AbstractFloat} <: CovEstimator{T}
-    n :: Int
-    μ :: Vector{T}
-    M :: Matrix{T}
-    δ :: Vector{T} # temporary
+function getest(nv::NaiveVar)
+    @assert nv.n >= 2 "Cannot estimate variance with only one sample"
+    return Statistics.var(nv.S)
 end
 
-function WelfordCov(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat}
-    d, = sz
-    WelfordCov(0, zeros(T, d), zeros(T, d, d), zeros(T, d))
-end
-
-function reset!(wc::WelfordCov{T}) where {T<:AbstractFloat}
-    wc.n = 0
-    wc.μ .= zero(T)
-    wc.M .= zero(T)
+# Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
+mutable struct WelfordVar{T<:AbstractVecOrMat{<:AbstractFloat}} <: VarEstimator{T}
+    n     :: Int
```

Comment:
Again, couldn't this be changed to
```suggestion
struct WelfordVar{S<:AbstractFloat,T<:AbstractVecOrMat{S}} <: VarEstimator{T}
```
?

If you need a consistent way for accessing the number of samples for different variance estimators, I'd suggest defining a function `nsamples(::WelfordVar) = ...` and so on instead of relying on a field `n`.

Reply:
It seems that for naive estimaters, there is no real need of accessing the number of samples. So I just remove that from naive ones and keep whatever we have for Welford.